src/apps/lwaftr/fragmentv6_hardened.lua

```diff
@@ -0,0 +1,299 @@
+
+ReassembleV6 = {}
+
+local function get_frag_len(frag)
+   local ipv6_payload_len = ehs + o_ipv6_payload_len
+   return C.ntohs(rd16(frag.data + ipv6_payload_len)) - ipv6_frag_header_size
+end
+
+local function get_frag_id(frag)
+   local o_id = ehs + ipv6_fixed_header_size + o_ipv6_frag_id
+   return C.ntohl(rd32(frag.data + o_id))
+end
+
+-- The least significant three bits are other information, but the
+-- offset is expressed in 8-octet units, so just mask them off.
+local function get_frag_start(frag)
+   local o_fstart = ehs + ipv6_fixed_header_size + o_ipv6_frag_offset
+   local raw_start = C.ntohs(rd16(frag.data + o_fstart))
+   local start = band(raw_start, 0xfff8)
+   return start
+end
+
+-- This is the 'M' bit of the IPv6 fragment header, in the
+-- least significant bit of the 4th byte
+local function is_last_fragment(frag)
+   local ipv6_frag_more_offset = ehs + ipv6_fixed_header_size + 3
+   return band(frag.data[ipv6_frag_more_offset], 1) == 0
+end
+
+local function get_key(fragment)
+   local key = scratch_fragkey
+   local o_src = ehs + o_ipv6_src_addr
+   local o_dst = ehs + o_ipv6_dst_addr
+   local o_id = ehs + ipv6_fixed_header_size + o_ipv6_frag_id
+   ffi.copy(key.src_addr, fragment.data + o_src, 16)
+   ffi.copy(key.dst_addr, fragment.data + o_dst, 16)
+   key.fragment_id = C.ntohl(rd32(fragment.data + o_id))
+   return key
+end
+
+local function free_reassembly_buf_and_pkt(pkt, frags_table)
+   local key = get_key(pkt)
+   frags_table:remove(key, false)
+   packet.free(pkt)
+end
+
+local function swap(array, i, j)
+   local tmp = array[j]
+   array[j] = array[i]
+   array[i] = tmp
+end
+
+-- This is an insertion sort, and only called on 2+ element arrays
+local function sort_array(array, last_index)
+   for i=0,last_index do
+      local j = i
+      while j > 0 and array[j-1] > array[j] do
+         swap(array, j, j-1)
+         j = j - 1
+      end
+   end
+end
+
+local function verify_valid_offsets(reassembly_buf)
+   if reassembly_buf.fragment_starts[0] ~= 0 then
+      return false
+   end
+   for i=1,reassembly_buf.fragment_count-1 do
+      if reassembly_buf.fragment_starts[i] ~= reassembly_buf.fragment_ends[i-1] then
+         return false
+      end
+   end
+   return true
+end
+
+local function reassembly_status(reassembly_buf)
+   if reassembly_buf.final_start == 0 then
+      return FRAGMENT_MISSING
+   end
+   if reassembly_buf.running_length ~= reassembly_buf.reassembly_packet.length then
+      return FRAGMENT_MISSING
+   end
+   if not verify_valid_offsets(reassembly_buf) then
+      return REASSEMBLY_INVALID
+   end
+   return REASSEMBLY_OK
+end
+
+local function attempt_reassembly(frags_table, reassembly_buf, fragment)
+   local reassembly_pkt = reassembly_buf.reassembly_packet
+   local frag_id = get_frag_id(fragment)
+   if frag_id ~= reassembly_buf.fragment_id then -- unreachable
+      assert(false, "Impossible case reached in v6 reassembly")
+      return REASSEMBLY_INVALID
+   end
```

Comment:
As assert is always false, the return statement is never reached. Why not?

```lua
assert(frag_id == reassembly_buf.fragment_id,  "Impossible case reached in v6 reassembly")
```
or

```lua
if frag_id ~= reassembly_buf.fragment_id then -- unreachable
   error("Impossible case reached in v6 reassembly")
end
```